### PR TITLE
vcs: add support for --max-instrs option

### PR DESCRIPTION
**Pull request by xiaokamikami**
Add a per-instruction counter to exit, primarily for checkpoint testing

### CODE REVIEW

File: src/test/csrc/difftest/difftest.cpp [hunk 1]
```diff
@@ -87,6 +87,20 @@ int difftest_step() {
...
+uint64_t difftest_commit_sum(char core_id) {
+  if (core_id < NUM_CORES)
+    return difftest[core_id]->get_instr_sum();
+  else 
+    return 0;
+}
+
+void difftest_commit_clean() {
+  for (uint32_t core_id = 0; core_id < NUM_CORES; core_id++)
+  {
```
Comment (poemonsense) [inline src/test/csrc/difftest/difftest.cpp:99]: Follow the same code style as existing code (to put `{` at the same line of `for`)

**Review by poemonsense — COMMENTED** [review]
let @klin02 have a review and test

File: src/test/csrc/difftest/difftest.cpp (hunk 1 above)
Comment (xiaokamikami) [reply src/test/csrc/difftest/difftest.cpp]: OK

File: src/test/csrc/vcs/vcs_main.cpp [hunk 2]
```diff
@@ -42,6 +43,10 @@ extern "C" void set_flash_bin(char *s) {
   strcpy(flash_bin_file, s);
 }
 
+extern "C" void set_max_instrs(long mc) {
```
Comment (klin02) [inline src/test/csrc/vcs/vcs_main.cpp:46]: It seems type `long` mismatchs DPI-C declare? Better use same type and more clear name then `mc`~

**Comment by klin02** [comment]
Better add exit Info like https://github.com/OpenXiangShan/difftest/blob/0330732fcdd22c7209866a3ae2213c8b29a4605b/src/test/vsrc/vcs/top.v#L199

File: src/test/csrc/difftest/difftest.cpp [hunk 3]
```diff
@@ -87,6 +87,19 @@ int difftest_step() {
...
 }
 
+uint64_t difftest_commit_sum(char core_id) {
+  if (core_id < NUM_CORES)
+    return difftest[core_id]->get_instr_sum();
+  else 
+    return 0;
+}
+
+void difftest_commit_clean() {
```
Comment (klin02) [inline src/test/csrc/difftest/difftest.cpp:97]: When will difftest_commit_clean be used? 
By the way, better use same code style as other difftest_xxx Func.
for (int i = 0; i < NUM_CORES; i++) ...

Same for other new function.

**Comment by klin02** [comment]
Does retire_instr_cntsum have any use other than controlling instruction overflow exit? Can trap->instrcnt be used as same role like emu?
https://github.com/OpenXiangShan/difftest/blob/0330732fcdd22c7209866a3ae2213c8b29a4605b/src/test/csrc/verilator/emu.cpp#L698
If some unused function is necessary for checkpoint. Maybe you can explain it in this PR?

File: src/test/csrc/vcs/vcs_main.cpp [hunk 4]
```diff
@@ -25,11 +25,13 @@
 #ifdef CONFIG_DIFFTEST_DEFERRED_RESULT
 #include "svdpi.h"
 #endif // CONFIG_DIFFTEST_DEFERRED_RESULT
+#define EXIT_MAX_INSTR 0xff
 
 static bool has_reset = false;
 static char bin_file[256] = "ram.bin";
 static char *flash_bin_file = NULL;
 static bool enable_difftest = true;
+static int max_instrs = 0;
```
Comment (klin02) [inline src/test/csrc/vcs/vcs_main.cpp:34]: use uint64_t? Line 47 arg is uint64_t, and trap->instrCnt is also uint64_t

File: src/test/csrc/vcs/vcs_main.cpp [hunk 5]
```diff
@@ -86,6 +92,14 @@ extern "C" int simv_step() {
     return trapCode + 1;
   }
 
+  if (max_instrs != 0) { // 0 for no limit
+    auto trap = difftest[0]->get_trap_event();
+    if(max_instrs < trap->instrCnt) {
+      eprintf(ANSI_COLOR_GREEN "checkpoint reach the operating limit exit\n" ANSI_COLOR_RESET);
```
Comment (klin02) [inline src/test/csrc/vcs/vcs_main.cpp:98]: Seems --max-instrs is independent to checkpoint. Simply `EXCEEDED MAX INSTR: %d` is enough

**Review by klin02 — CHANGES_REQUESTED** [review]

**Comment by poemonsense** [comment]
In Linux, the return code of a process indicates the state of the process. See https://www.agileconnection.com/article/overview-linux-exit-codes

Return code 0 means the program exits normally. For max-instrs, the simulation should return 0 as well. This is also the normal behavior in the Verilator simulation as shown in https://github.com/OpenXiangShan/difftest/blob/master/src/test/csrc/common/main.cpp#L45 and https://github.com/OpenXiangShan/difftest/blob/master/src/test/csrc/verilator/emu.h#L138

Please do not use 0xff as the return code

**Comment by xiaokamikami** [comment]
> In Linux, the return code of a process indicates the state of the process. See https://www.agileconnection.com/article/overview-linux-exit-codes
> 
> Return code 0 means the program exits normally. For max-instrs, the simulation should return 0 as well. This is also the normal behavior in the Verilator simulation as shown in https://github.com/OpenXiangShan/difftest/blob/master/src/test/csrc/common/main.cpp#L45 and https://github.com/OpenXiangShan/difftest/blob/master/src/test/csrc/verilator/emu.h#L138
> 
> Please do not use 0xff as the return code

In the next version I need to determine if it's max-instrs and run a separate handler

**Comment by poemonsense** [comment]
> > In Linux, the return code of a process indicates the state of the process. See https://www.agileconnection.com/article/overview-linux-exit-codes
> > Return code 0 means the program exits normally. For max-instrs, the simulation should return 0 as well. This is also the normal behavior in the Verilator simulation as shown in https://github.com/OpenXiangShan/difftest/blob/master/src/test/csrc/common/main.cpp#L45 and https://github.com/OpenXiangShan/difftest/blob/master/src/test/csrc/verilator/emu.h#L138
> > Please do not use 0xff as the return code
> 
> In the next version I need to determine if it's max-instrs and run a separate handler

OK. It's only for internal usage

File: src/test/csrc/vcs/vcs_main.cpp (hunk 5 above)
Comment (xiaokamikami) [reply src/test/csrc/vcs/vcs_main.cpp]: fix

File: src/test/csrc/vcs/vcs_main.cpp (hunk 4 above)
Comment (xiaokamikami) [reply src/test/csrc/vcs/vcs_main.cpp]: fix

**Review by klin02 — APPROVED** [review]

**Pull request merged**